subsys/net/ip/net_mgmt.c

```diff
@@ -167,9 +167,9 @@ static inline void mgmt_run_callbacks(struct mgmt_event_entry *mgmt_event)
 		NET_MGMT_GET_COMMAND(mgmt_event->event));
 
 	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&event_callbacks, cb, tmp, node) {
-		if (!(NET_MGMT_GET_LAYER(mgmt_event->event) ==
+		if (!(NET_MGMT_GET_LAYER(mgmt_event->event) &
```

Comment:
Nack to that. The mask is only for command. Layer and layer code have to be an exact match.

Reply:
> The mask is only for command. Layer and layer code have to be an exact match.

Is documented properly somewhere? On my memory, this causes a stream of confusion and bugs in the code. Perhaps, besides documentation, we also should provide helper macros to do the matching, and require their usage.

Reply:
#16133

Reply:
I think we need to change this, how this works now is just counter intuitive. For example I was trying to catch these events `NET_EVENT_IPV6_ADDR_DEL |  NET_EVENT_IPV4_ADDR_ADD | NET_EVENT_IPV4_ADDR_DEL` and it was not possible to catch that. It makes no sense to have multiple callbacks just to catch that.

Reply:
no, this won't work. It actually worked like that before and triggered bugs, too many false positive cb calls etc... And in the end, in the code side: it's ugly to mix layers and all. Let's keep it as it is.

Reply:
We need to add more unit tests so the things will work as expected. This is constant battle to keep things working if we change things and do not have proper tests in place.

As I said earlier, as a net-mgmt library user I am expecting that I am able to use the sync version of the net_mgmt API to catch things properly. Currently this is not possible if I need to have multiple masks defined to get the desired events like I mentioned above.

Reply:
sync version? you mean socket version? If the API is a 1:1 on core net_mgmt I don't see any problem.

Now if you want to modify the API that's another story. And that patch is not going to work properly for what you are trying to do.

Reply:
> sync version? you mean socket version? If the API is a 1:1 on core net_mgmt I don't see any problem.

By sync version I am meaning `net_mgmt_event_wait_on_iface()` and `net_mgmt_event_wait_on_iface()`

> Now if you want to modify the API that's another story. And that patch is not going to work properly for what you are trying to do.

No, I do not want to change the API unless we have to. I am expecting that I can call either of the above functions with this mask `NET_EVENT_IPV6_ADDR_DEL | NET_EVENT_IPV4_ADDR_ADD | NET_EVENT_IPV4_ADDR_DEL` as currently it does not work.

Reply:
It is an API change, since layer and layer code would be then interpreted as a mask. And your patch is not going to work: layer and layer code have overlapping bits. (false positives, etc...)
All of this has been discussed long time ago, and a fix would require to rework the bits fields, there was even discussion back then to extend the whole to 64 bits etc...

Not that it is not possible, but it's a different thing than just exposing net_mgmt through socket.